keps/0021-kudo-upgrade.md

```diff
@@ -0,0 +1,271 @@
+the least likely to change, but probably the most specific. If we make changes here, we need to implement custom
+migration code.
+
+- Current prereqs
+  - Namespace
+  - Service Account
+  - Role Bindings
+  - Secret
+
+#### Proposal for update process 
+Integrated into `kudo init --upgrade`
+
+Write specific migration code that targets a KUDO manager version range and executes manual migration steps. 
+
+Each migration should have a validate-step that checks if the migration is possible.
+
+**Alternative update process**
+- Can we just delete them all and reinstall them? Probably not
+
+### KUDO Manager
+Expected update frequency: High  
+Versioned: Yes
+
+The KUDO Manager is defined by an image version in a deployment set. To update, the deployment must be updated. The 
+manager is closely tied to the CRDs, but not to the CLI. When CRDs are updated, the Manager will most likely also
+need to be updated. 
+
+#### Proposal for update process
+Integrated into `kudo init --upgrade`
+
+- Use semantic versioning for the manager binary
+  - ToBeDiscussed: Use different versioning for Manager and CLI?
+- The manager needs to be shut down before updating the CRDs
+  - As soon as the new CRD is installed and marked as "stored"
+  - Without WebHook Conversion, even if the new CRD version is "served", the old content is returned without any conversion
+- Question: Do we need to ensure that the manager is not doing meaningful work at the moment, or can we just update the deployment?
+
+### CRDs
+Expected update frequency: Medium  
+Versioned: Yes, with a CRD-Version
+
+The CRDs are used to store installed operators and running instances. New features will regularly require us to add new
+fields. 
+- Existing CRs need to be migrated to new versions
+- K8s CRD support
+  - Which K8s version do we want/need to support?
+    - MultiVersion is supported since 1.11 (manual conversion)
+    - WebHook conversion since 1.16
+  - WebHook conversion would allow us to transparently switch to a new CRD version without manually migrating all existing CRs
+- CRD versioning
+  - Currently we have one global version for all CRDs
+  - Alternatively, we could have a one version per CRD
+  - Single version should be ok, in the worst case we have empty migrations
+  - One version for each CRD would require complex compatibility matrix
+
+#### Proposal for update process
+Integrated into `kudo init --upgrade`
+
+To support older K8s versions, no WebHook conversion is used. We only serve a single CRD-Version and migrate existing CRs in the update process
+
+### KUDO CLI
+Expected update frequency: High
+
+KUDO CLI is the command line tool to manage KUDO. It will be often updated to add new features and fix bugs. It needs
+to be in sync with the installed CRDs, as it's writing them directly with the K8s extension API.
+
+- Do we allow an older KUDO CLI to be used with a newer KUDO installation?
+
+#### Proposal for update process
+User has to download newest KUDO version.
```

Comment:
yeah I think replacing the binary or dowing upgrade via brew is good enough here 👍

Reply:
Yeah, i think this is the most simple part of the update process :)